### DISCUSSION TITLE
redirect dose't work & ssr dose't work

### DISCUSSION THREAD

**Discussion by dantine123** [nuxt/nuxt#9525] [Questions (Nuxt 2)]
hello,
my first problem is redirect dose't work in middleware, 
console.log('redirect') is ok, 
but 'redirect('/login')' cant work, nuxt version is 2.15.7
auth.js:
![微信截图_20210706102005](https://user-images.githubusercontent.com/32285465/124532825-d6a14380-de43-11eb-883f-9bf59a2628e2.png)
userCenter.vue
![微信截图_20210706101819](https://user-images.githubusercontent.com/32285465/124532834-dbfe8e00-de43-11eb-84b5-338b992e93d1.png)

my second problem is when i  use nuxt-link or this.$router.push to jump to homepage, homepage ssr (i used asyncData) dont work。 instead， the requests in  asyncData  runed at frontend by ajax, like
![微信截图_20210706103404](https://user-images.githubusercontent.com/32285465/124534624-39480e80-de47-11eb-9dc8-5be97695c552.png)
how to fixed?thxs~!

**Comment by kissu** [comment]
What are your `target` and `ssr` values in `nuxt.config.js`?

With `ssr: false` and `target: static`
![image](https://user-images.githubusercontent.com/5133074/124609341-d3bc4800-de6f-11eb-9f56-72a87861ffc4.png)
With `ssr: true` and `target: static`
![image](https://user-images.githubusercontent.com/5133074/124609493-f2bada00-de6f-11eb-9f0e-838497583f76.png)
There is a double console.log because one is done on the server and the other on the client.

With `ssr: true` and `target: server`, the code is not even reaching the browser because it is returned when generating the page (with the `redirect`)
![image](https://user-images.githubusercontent.com/5133074/124609877-51805380-de70-11eb-898b-0f19dd73414f.png)
Hence, the following error in the browser
![image](https://user-images.githubusercontent.com/5133074/124609922-5cd37f00-de70-11eb-8010-57e67e2a3237.png)

This to run this lighter snippet of code as a middleware and understand what is happening with various configurations
```js
export default ({ app, redirect }) => {
  console.log('process and cookies', process.server && !!app.$cookies.get('token'))

  console.log('server', process.server)
  console.log('client', process.client)
  console.log('static', process.static)
  if (process.server) {
    console.log('redirect will be called shortly')
    redirect('/login')
  }
}
```

So, every behavior is working as expected IMO.
Also, please check this page of the documentation: https://nuxtjs.org/docs/2.x/concepts/context-helpers#process-helpers

PS: for people who are willing to try this, OP is using [cookie-universal-nuxt](https://www.npmjs.com/package/cookie-universal-nuxt) when he uses `app.$cookies.get`.

---
The second part of your question is probably working as expected too, since your SPA is following what you've asked it to do (it moves thanks to `nuxt-router`, aka a browser navigation **only**).
The AJAX call is also totally legit here.
